ManalHamdi/assignment2/Assignment2_q2.cc

```diff
@@ -0,0 +1,101 @@
+#include <iostream>
+#include <queue>
+#include <vector>
+
+struct Node {
+	int data;
+	Node* parent;
+	Node* left_child;
+	Node* right_child; 
+};
+class Tree {
+public:
+	Node* root;
+	Tree() {
+		root = nullptr;
+	}
+	Node* CreateNode(int data) {
+		Node* temp = new Node;
```

Comment:
This line is problematic because you're leaking memory here.

Consider if you have a statement:

```
Node* CreateNode(int data) {
    Node temp;
    ....
```

The `temp` node you are creating here is created in 'local' memory, on the function's stack. This means that the variable will be deleted when you exist from the function. If you try to return it from the function with:

 `return &temp;`

you will get a segmentation fault because by the time you return from a function, 'temp' no longer exists.

In this function, you want to return a pointer to a `Node`, so you need to create it and be able to pass it back to the caller. This can be done with what you're doing, by using the `new` keyword:

`Node* temp = new Node;`

When you use `new`, the variable you're creating lives on the program's heap and will never be deleted automatically. It is therefore the responsibility of whoever creates a `new` object to delete it when it is no longer needed.

So for each node you create you must also do:

`delete temp;`

to delete it (in your case it will be in the last line of your main function).

C++ makes this somewhat easier by providing the concept of [destructors](https://www.geeksforgeeks.org/destructors-c/). To release memory for all of the nodes in your tree, you just need to add the following to your class definition:

```
class Tree {
    public:
        ~Tree(){
           delete root;
        }
}
```

But this is not enough! `struct Node` is actually the one that holds all of the pointers to the children and the parent nodes, so if you want your tree to be deleted recursively (i.e. *all* of the nodes are deleted instead of just stopping at the root), you have to add a destructor to `Node` too:

```
struct Node {
    ....
    ~Node() {
        // Here you don't strictly need to delete parent, because if you only use 'Node' with the Tree class,
        // you're already deleting the parent when you call 'delete root' in '~Tree()'.
        delete right_child;
        delete left_child;
    }
```
Now, when your program exits `~Tree()` will be called automatically and `delete root` will call `~Node()`. Each `delete` statement in `~Node() ` will then call `Node` destructors on each of the children recursively.

In the Hangout, I also mentioned [std::unique_ptr](http://www.drdobbs.com/cpp/c11-uniqueptr/240002708), which makes memory management even simpler :) But for now, please add destructors to your object definitions to hopefully understand these important concepts better and we can try unique_ptrs in the next assignment :)

Reply:
Also `CreateNode` should be marked as a [static](https://www.geeksforgeeks.org/static-keyword-cpp/) function, because you're not modifying object state in it.